kubespawner/clients.py

```diff
@@ -41,8 +44,29 @@ def shared_client(ClientType, *args, **kwargs):
         # Kubernetes client configuration is handled globally
         # in kubernetes.py and is already called in spawner.py
         # or proxy.py prior to a shared_client being instantiated
-        Client = getattr(kubernetes.client, ClientType)
+        Client = getattr(kubernetes_asyncio.client, ClientType)
         client = Client(*args, **kwargs)
         # cache weakref so that clients can be garbage collected
         _client_cache[cache_key] = weakref.ref(client)
+
     return client
+
+
+async def load_config(caller=None):
+    try:
+        kubernetes_asyncio.config.load_incluster_config()
+    except kubernetes_asyncio.config.ConfigException:
+        await kubernetes_asyncio.config.load_kube_config()
+    # The ca_cert/k8s_api_host is set via traitlets on the objects
+    # that call load_config(); thus, they may pass a reference to themselves
+    # into load_config in order to set ca_cert/k8s_api_host.
+
+    if caller:
+        if hasattr(caller, "k8s_api_ssl_ca_cert") and caller.k8s_api_ssl_ca_cert:
+            global_conf = Configuration.get_default_copy()
+            global_conf.ssl_ca_cert = caller.k8s_api_ssl_ca_cert
+            Configuration.set_default(global_conf)
+        if hasattr(caller, "k8s_api_host") and caller.k8s_api_host:
+            global_conf = Configuration.get_default_copy()
+            global_conf.host = caller.k8s_api_host
+            Configuration.set_default(global_conf)
```

Comment:
I figure `caller` is better made not optional. Having it optional made the code become more complicated, and made me think that perhaps this function was actively used in a way were it was optional - which it wasn't.

```suggestion
async def load_config(caller):
    """
    Loads configuration for the Python client we use to communicate with a k8s
    api-server, and tweaks that configuration based on specific settings on the
    passed caller object.
    """
    try:
        kubernetes_asyncio.config.load_incluster_config()
    except kubernetes_asyncio.config.ConfigException:
        await kubernetes_asyncio.config.load_kube_config()

    if caller.k8s_api_ssl_ca_cert:
        global_conf = Configuration.get_default_copy()
        global_conf.ssl_ca_cert = caller.k8s_api_ssl_ca_cert
        Configuration.set_default(global_conf)
    if caller.k8s_api_host:
        global_conf = Configuration.get_default_copy()
        global_conf.host = caller.k8s_api_host
        Configuration.set_default(global_conf)
```

Reply:
I don't see how we avoid calling it from the Reflector, unless we make the Reflector require a parent object.

That might be the right thing to do...or we could just add those attributes to the Reflector too.

Reply:
And parent is always specified, so all we need to do is check that it's not None.